src/algebra/lie/basic.lean

```diff
@@ -1267,34 +1285,235 @@ end lie_submodule
+def ker : lie_ideal R L := lie_ideal.comap f ⊥
+
+/-- The range of a morphism of Lie algebras as an ideal in the codomain. -/
+def ideal_range : lie_ideal R L' := lie_ideal.map f ⊤
+
+/-- The condition that the image of a morphism of Lie algebras is an ideal. -/
+def is_ideal_morphism : Prop := (f.ideal_range : lie_subalgebra R L') = f.range
+
+@[simp] lemma is_ideal_morphism_def :
+  f.is_ideal_morphism ↔ (f.ideal_range : lie_subalgebra R L') = f.range := iff.rfl
+
+lemma range_subset_ideal_range : (f.range : set L') ⊆ f.ideal_range := lie_submodule.subset_lie_span
+
+lemma map_le_ideal_range : I.map f ≤ f.ideal_range := lie_ideal.map_mono le_top
+
+lemma ker_le_comap : f.ker ≤ J.comap f := lie_ideal.comap_mono bot_le
+
+@[simp] lemma mem_ker {x : L} (h : x ∈ ker f) : f x = 0 :=
+begin
+  change x ∈ lie_ideal.comap f ⊥ at h,
+  rw [lie_ideal.mem_comap, lie_submodule.mem_bot] at h, exact h,
+end
+
+lemma mem_ideal_range {x : L} : f x ∈ ideal_range f := lie_ideal.mem_map (lie_submodule.mem_top x)
+
+@[simp] lemma mem_ideal_range_iff (h : is_ideal_morphism f) {y : L'} :
+  y ∈ ideal_range f ↔ ∃ (x : L), f x = y :=
+begin
+  rw f.is_ideal_morphism_def at h,
+  rw [← lie_submodule.mem_coe, ← lie_ideal.coe_to_subalgebra, h, f.range_coe, set.mem_range],
+end
+
 
-lemma map_le_iff_le_comap {f : L →ₗ⁅R⁆ L'} {I : lie_ideal R L} {I' : lie_ideal R L'} :
-  map f I ≤ I' ↔ I ≤ comap f I' :=
-by { erw lie_submodule.lie_span_le, exact set.image_subset_iff, }
+@[simp] lemma ker_coe_submodule : (ker f : submodule R L) = (f : L →ₗ[R] L').ker := rfl
 
-lemma gc_map_comap (f : L →ₗ⁅R⁆ L') : galois_connection (map f) (comap f) :=
+end lie_algebra.morphism
+
+namespace lie_ideal
+
+variables {f : L →ₗ⁅R⁆ L'} {I : lie_ideal R L} {J : lie_ideal R L'}
+
+lemma map_sup_ker_eq_map : lie_ideal.map f (I ⊔ f.ker) = lie_ideal.map f I :=
+begin
+  suffices : lie_ideal.map f (I ⊔ f.ker) ≤ lie_ideal.map f I,
+  { exact le_antisymm this (lie_ideal.map_mono le_sup_left), },
+  apply lie_submodule.lie_span_mono,
+  rintros x ⟨y, hy₁, hy₂⟩, rw ← hy₂,
+  erw lie_submodule.mem_sup at hy₁, obtain ⟨z₁, hz₁, z₂, hz₂, hy⟩ := hy₁, rw ← hy,
+  rw [f.map_add, f.mem_ker hz₂, add_zero], exact ⟨z₁, hz₁, rfl⟩,
+end
+
+@[simp] lemma map_comap_eq (h : f.is_ideal_morphism) : map f (comap f J) = f.ideal_range ⊓ J :=
+begin
+  apply le_antisymm,
+  { rw le_inf_iff, exact ⟨f.map_le_ideal_range _, map_comap_le⟩, },
+  { rw f.is_ideal_morphism_def at h,
+    rw [lie_submodule.le_def, lie_submodule.inf_coe, ← coe_to_subalgebra, h],
+    rintros y ⟨⟨x, h₁, h₂⟩, h₃⟩, rw ← h₂ at h₃ ⊢, exact mem_map h₃, },
+end
+
+@[simp] lemma comap_map_eq (h : ↑(map f I) = f '' I) : comap f (map f I) = I ⊔ f.ker :=
+by rw [← lie_submodule.coe_to_submodule_eq_iff, comap_coe_submodule, I.map_coe_submodule f h,
+  lie_submodule.sup_coe_to_submodule, f.ker_coe_submodule, linear_map.comap_map_eq]
+
+variables (f I J)
+
+lemma gc_map_comap : galois_connection (map f) (comap f) :=
```

Comment:
I would move this up ~100 lines, and put it right after `map_le_iff_le_comap`.
Otherwise, LGTM.
It will be useful for proving that `map` and `comap` preserve some order-related stuff.

bors d+